File: mcp-spike/core-app/src/main/java/org/slizaa/mcp/core/rest/GraphController.java
package org.slizaa.mcp.core.rest;

import org.slizaa.mcp.core.DetailDependenciesMcpTool;
import org.slizaa.mcp.core.DiscoveryMcpTools;
import org.slizaa.mcp.core.FieldDetailsMcpTool;
import org.slizaa.mcp.core.ListFieldsMcpTool;
import org.slizaa.mcp.core.ListMethodsMcpTool;
import org.slizaa.mcp.core.MethodDetailsMcpTool;
import org.slizaa.mcp.core.PairwiseDependencyMcpTools;
import org.slizaa.mcp.core.ReachabilityMcpTools;
import org.slizaa.mcp.core.ScopeDependencyMcpTools;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api")
public class GraphController {

    private final DiscoveryMcpTools discoveryTools;
    private final PairwiseDependencyMcpTools pairwiseTools;
    private final ScopeDependencyMcpTools scopeTools;
    private final ReachabilityMcpTools reachabilityTools;
    private final ListMethodsMcpTool listMethodsTool;
    private final ListFieldsMcpTool listFieldsTool;
    private final DetailDependenciesMcpTool detailDependenciesTool;
    private final MethodDetailsMcpTool methodDetailsTool;
    private final FieldDetailsMcpTool fieldDetailsTool;

    public GraphController(DiscoveryMcpTools discoveryTools,
                           PairwiseDependencyMcpTools pairwiseTools,
                           ScopeDependencyMcpTools scopeTools,
                           ReachabilityMcpTools reachabilityTools,
                           ListMethodsMcpTool listMethodsTool,
                           ListFieldsMcpTool listFieldsTool,
                           DetailDependenciesMcpTool detailDependenciesTool,
                           MethodDetailsMcpTool methodDetailsTool,
                           FieldDetailsMcpTool fieldDetailsTool) {
        this.discoveryTools = discoveryTools;
        this.pairwiseTools = pairwiseTools;
        this.scopeTools = scopeTools;
        this.reachabilityTools = reachabilityTools;
        this.listMethodsTool = listMethodsTool;
        this.listFieldsTool = listFieldsTool;
        this.detailDependenciesTool = detailDependenciesTool;
        this.methodDetailsTool = methodDetailsTool;
        this.fieldDetailsTool = fieldDetailsTool;
    }

    @GetMapping("/find-node")
    public List<Map<String, Object>> findNode(
            @RequestParam String query,
            @RequestParam(required = false) String kind,
            @RequestParam(required = false) Integer limit) {
        return discoveryTools.findNode(query, kind, limit);
    }

    @GetMapping("/list-children")
    public List<Map<String, Object>> listChildren(
            @RequestParam(required = false) Long nodeId,
            @RequestParam(required = false) Integer limit) {
        return discoveryTools.listChildren(nodeId, limit);
    }

    @GetMapping("/list-descendants")
    public Map<String, Object> listDescendants(
            @RequestParam long rootId,
            @RequestParam(required = false) List<String> kindFilter,
            @RequestParam(required = false) List<String> excludeKindFilter,
            @RequestParam(required = false) Integer limit) {
        return discoveryTools.listDescendants(rootId, kindFilter, excludeKindFilter, limit);
    }

    @GetMapping("/dependency-between")
    public Map<String, Object> dependencyBetween(
            @RequestParam long fromId,
            @RequestParam long toId) {
        return pairwiseTools.dependencyBetween(fromId, toId);
    }

    @GetMapping("/aggregated-outgoing")
    public Map<String, Object> aggregatedOutgoing(
            @RequestParam long sourceId,
            @RequestParam(required = false) Long targetScopeId,
            @RequestParam(required = false) Integer limit) {
        return scopeTools.aggregatedOutgoing(sourceId, targetScopeId, limit);
    }

    @GetMapping("/aggregated-incoming")
    public Map<String, Object> aggregatedIncoming(
            @RequestParam long targetId,
            @RequestParam(required = false) Long sourceScopeId,
            @RequestParam(required = false) Integer limit) {
        return scopeTools.aggregatedIncoming(targetId, sourceScopeId, limit);
    }

    @GetMapping("/outgoing-core-dependencies")
    public Map<String, Object> outgoingCoreDependencies(
            @RequestParam long fromId,
            @RequestParam long toId,
            @RequestParam(required = false) Integer limit) {
        return scopeTools.outgoingCoreDependencies(fromId, toId, limit);
    }

    @GetMapping("/incoming-core-dependencies")
    public Map<String, Object> incomingCoreDependencies(
            @RequestParam long toId,
            @RequestParam long fromId,
            @RequestParam(required = false) Integer limit) {
        return scopeTools.incomingCoreDependencies(toId, fromId, limit);
    }

    @GetMapping("/describe-graph")
    public Map<String, Object> describeGraph(
            @RequestParam(required = false) Long scopeId) {
        return discoveryTools.describeGraph(scopeId);
    }

    @GetMapping("/find-dependency-path")
    public Map<String, Object> findDependencyPath(
            @RequestParam long fromId,
            @RequestParam long toId,
            @RequestParam(required = false) Integer maxLength) {
        return reachabilityTools.findDependencyPath(fromId, toId, maxLength);
    }

    @GetMapping("/pairwise-dependencies")
    public Map<String, Object> pairwiseDependencies(
            @RequestParam List<Long> nodeIds,
            @RequestParam(required = false) Boolean includeSelfLoops) {
        return reachabilityTools.pairwiseDependencies(nodeIds, includeSelfLoops);
    }

    @GetMapping("/affected-by")
    public Map<String, Object> affectedBy(
            @RequestParam long sourceId,
            @RequestParam(required = false) Integer maxDepth,
            @RequestParam(required = false) Long groupingScopeId,
            @RequestParam(required = false) Integer topN) {
        return reachabilityTools.affectedBy(sourceId, maxDepth, groupingScopeId, topN);
    }

    @GetMapping("/outgoing-to")
    public Map<String, Object> outgoingTo(
            @RequestParam long sourceId,
            @RequestParam List<Long> targetIds,
            @RequestParam(required = false) Boolean includeMissing) {
        return pairwiseTools.outgoingTo(sourceId, targetIds, includeMissing);
    }

    @GetMapping("/incoming-from")
    public Map<String, Object> incomingFrom(
            @RequestParam long targetId,
            @RequestParam List<Long> sourceIds,
            @RequestParam(required = false) Boolean includeMissing) {
        return pairwiseTools.incomingFrom(targetId, sourceIds, includeMissing);
    }

    // --- Detail-level tools ---

    @GetMapping("/list-methods")
    public Map<String, Object> listMethods(
            @RequestParam long typeId,
            @RequestParam(required = false) String namePattern,
            @RequestParam(required = false) List<String> modifierFilter,
            @RequestParam(required = false) Boolean includeInherited,
            @RequestParam(required = false) Integer limit) {
        return listMethodsTool.listMethods(typeId, namePattern, modifierFilter, includeInherited, limit);
    }

    @GetMapping("/list-fields")
    public Map<String, Object> listFields(
            @RequestParam long typeId,
            @RequestParam(required = false) String namePattern,
            @RequestParam(required = false) List<String> modifierFilter,
            @RequestParam(required = false) Boolean includeInherited,
            @RequestParam(required = false) Integer limit) {
        return listFieldsTool.listFields(typeId, namePattern, modifierFilter, includeInherited, limit);
    }

    @GetMapping("/detail-dependencies")
    public Map<String, Object> detailDependencies(
            @RequestParam long fromId,
            @RequestParam long toId,
            @RequestParam(required = false) String relationship,
            @RequestParam(required = false) Integer limit) {
        return detailDependenciesTool.detailDependencies(fromId, toId, relationship, limit);
    }

    @GetMapping("/method-details")
    public Map<String, Object> methodDetails(@RequestParam long methodId) {
        return methodDetailsTool.methodDetails(methodId);
    }

    @GetMapping("/field-details")
    public Map<String, Object> fieldDetails(@RequestParam long fieldId) {
        return fieldDetailsTool.fieldDetails(fieldId);
    }
}

File: mcp-spike/core-app/src/main/kotlin/org/slizaa/mcp/core/rest/GraphController.kt
package org.slizaa.mcp.core.rest

import org.slizaa.mcp.core.mcp.detail.DetailDependenciesMcpTool
import org.slizaa.mcp.core.mcp.navigation.ListDescendantsTool
import org.slizaa.mcp.core.mcp.detail.FieldDetailsMcpTool
import org.slizaa.mcp.core.mcp.detail.ListFieldsMcpTool
import org.slizaa.mcp.core.mcp.detail.ListMethodsMcpTool
import org.slizaa.mcp.core.mcp.detail.MethodDetailsMcpTool
import org.slizaa.mcp.core.mcp.navigation.FindNodeTool
import org.slizaa.mcp.core.mcp.navigation.GraphOverviewTool
import org.slizaa.mcp.core.mcp.navigation.ListChildrenTool
import org.slizaa.mcp.core.mcp.pairwisedependency.PairwiseDependencyMcpTools
import org.slizaa.mcp.core.mcp.reachability.ReachabilityMcpTools
import org.slizaa.mcp.core.mcp.scopedependency.ScopeDependencyMcpTools
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api")
class GraphController(
    private val findNodeTool: FindNodeTool,
    private val graphOverviewTool: GraphOverviewTool,
    private val listChildrenTool: ListChildrenTool,
    private val listDescendantsTool: ListDescendantsTool,
    private val pairwiseTools: PairwiseDependencyMcpTools,
    private val scopeTools: ScopeDependencyMcpTools,
    private val reachabilityTools: ReachabilityMcpTools,
    private val listMethodsTool: ListMethodsMcpTool,
    private val listFieldsTool: ListFieldsMcpTool,
    private val detailDependenciesTool: DetailDependenciesMcpTool,
    private val methodDetailsTool: MethodDetailsMcpTool,
    private val fieldDetailsTool: FieldDetailsMcpTool
) {

    @GetMapping("/find-node")
    fun findNode(
        @RequestParam name: String,
        @RequestParam(required = false) kindFilter: List<String>?
    ): Map<String, *> = findNodeTool.findNode(name, kindFilter)

    @GetMapping("/list-children")
    fun listChildren(
        @RequestParam nodeId: Long,
        @RequestParam(required = false) kindFilter: List<String>?,
        @RequestParam(required = false) namePattern: String?,
        @RequestParam(required = false) modifierFilter: List<String>?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = listChildrenTool.listChildren(nodeId, kindFilter, namePattern, modifierFilter, limit)

    @GetMapping("/list-descendants")
    fun listDescendants(
        @RequestParam nodeId: Long,
        @RequestParam(required = false) kindFilter: List<String>?,
        @RequestParam(required = false) namePattern: String?,
        @RequestParam(required = false) modifierFilter: List<String>?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = listDescendantsTool.listDescendants(nodeId, kindFilter, namePattern, modifierFilter, limit)

    @GetMapping("/dependency-between")
    fun dependencyBetween(
        @RequestParam fromId: Long,
        @RequestParam toId: Long
    ): Map<String, Any?> = pairwiseTools.dependencyBetween(fromId, toId)

    @GetMapping("/aggregated-outgoing")
    fun aggregatedOutgoing(
        @RequestParam sourceId: Long,
        @RequestParam(required = false) targetScopeId: Long?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = scopeTools.aggregatedOutgoing(sourceId, targetScopeId, limit)

    @GetMapping("/aggregated-incoming")
    fun aggregatedIncoming(
        @RequestParam targetId: Long,
        @RequestParam(required = false) sourceScopeId: Long?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = scopeTools.aggregatedIncoming(targetId, sourceScopeId, limit)

    @GetMapping("/outgoing-core-dependencies")
    fun outgoingCoreDependencies(
        @RequestParam fromId: Long,
        @RequestParam toId: Long,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = scopeTools.outgoingCoreDependencies(fromId, toId, limit)

    @GetMapping("/incoming-core-dependencies")
    fun incomingCoreDependencies(
        @RequestParam toId: Long,
        @RequestParam fromId: Long,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = scopeTools.incomingCoreDependencies(toId, fromId, limit)

    @GetMapping("/graph-overview")
    fun graphOverview(): Map<String, Any?> = graphOverviewTool.graphOverview()

    @GetMapping("/find-dependency-path")
    fun findDependencyPath(
        @RequestParam fromId: Long,
        @RequestParam toId: Long,
        @RequestParam(required = false) maxLength: Int?
    ): Map<String, Any?> = reachabilityTools.findDependencyPath(fromId, toId, maxLength)

    @GetMapping("/pairwise-dependencies")
    fun pairwiseDependencies(
        @RequestParam nodeIds: List<Long>,
        @RequestParam(required = false) includeSelfLoops: Boolean?
    ): Map<String, Any?> = reachabilityTools.pairwiseDependencies(nodeIds, includeSelfLoops)

    @GetMapping("/affected-by")
    fun affectedBy(
        @RequestParam sourceId: Long,
        @RequestParam(required = false) maxDepth: Int?,
        @RequestParam(required = false) groupingScopeId: Long?,
        @RequestParam(required = false) topN: Int?
    ): Map<String, Any?> = reachabilityTools.affectedBy(sourceId, maxDepth, groupingScopeId, topN)

    @GetMapping("/outgoing-to")
    fun outgoingTo(
        @RequestParam sourceId: Long,
        @RequestParam targetIds: List<Long>,
        @RequestParam(required = false) includeMissing: Boolean?
    ): Map<String, Any?> = pairwiseTools.outgoingTo(sourceId, targetIds, includeMissing)

    @GetMapping("/incoming-from")
    fun incomingFrom(
        @RequestParam targetId: Long,
        @RequestParam sourceIds: List<Long>,
        @RequestParam(required = false) includeMissing: Boolean?
    ): Map<String, Any?> = pairwiseTools.incomingFrom(targetId, sourceIds, includeMissing)

    // --- Detail-level tools ---

    @GetMapping("/list-methods")
    fun listMethods(
        @RequestParam typeId: Long,
        @RequestParam(required = false) namePattern: String?,
        @RequestParam(required = false) modifierFilter: List<String>?,
        @RequestParam(required = false) includeInherited: Boolean?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = listMethodsTool.listMethods(typeId, namePattern, modifierFilter, includeInherited, limit)

    @GetMapping("/list-fields")
    fun listFields(
        @RequestParam typeId: Long,
        @RequestParam(required = false) namePattern: String?,
        @RequestParam(required = false) modifierFilter: List<String>?,
        @RequestParam(required = false) includeInherited: Boolean?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = listFieldsTool.listFields(typeId, namePattern, modifierFilter, includeInherited, limit)

    @GetMapping("/detail-dependencies")
    fun detailDependencies(
        @RequestParam fromId: Long,
        @RequestParam toId: Long,
        @RequestParam(required = false) relationship: String?,
        @RequestParam(required = false) limit: Int?
    ): Map<String, Any?> = detailDependenciesTool.detailDependencies(fromId, toId, relationship, limit)

    @GetMapping("/method-details")
    fun methodDetails(@RequestParam methodId: Long): Map<String, Any?> =
        methodDetailsTool.methodDetails(methodId)

    @GetMapping("/field-details")
    fun fieldDetails(@RequestParam fieldId: Long): Map<String, Any?> =
        fieldDetailsTool.fieldDetails(fieldId)
}
